site/specification/ORCv1.md

```diff
@@ -369,6 +392,174 @@ message Metadata {
+by the Hadoop or Ranger Key Management Server (KMS). The encrypted
+local keys are stored in the file footer's StripeInformation.
+
+```
+message Encryption {
+  // all of the masks used in this file
+  repeated DataMask mask = 1;
+  // all of the keys used in this file
+  repeated EncryptionKey key = 2;
+  // The encrypted variants.
+  // Readers should prefer the first variant that the user has access to
+  // the corresponding key. If they don't have access to any of the keys,
+  // they should get the unencrypted data.
+  repeated EncryptionVariant variants = 3;
+}
+```
+
+## Data Masks
+
+The user's data is statically masked before writing the unencrypted
+variant. Because the masking was done statically when the file was
+written, the information about the masking is just informational.
+
+The three standard masks are:
+
+* nullify - all values become null
+* redact - replace characters with constants such as X or 9
+* sha256 - replace string with the SHA 256 of the value
+
+The default is nullify, but masks may be defined by the user. Masks
+are not allowed to change the type of the column, just the values.
+
+```
+message DataMask {
+  // the kind of masking, which may include third party masks
+  optional string name = 1;
+  // parameters for the mask
+  repeated string maskParameters = 2;
+  // the unencrypted column roots this mask was applied to
+  repeated uint32 columns = 3 [packed = true];
+}
+```
+
+## Encryption Keys
+
+In addition to the encrypted local keys, which are stored in the
+footer's StripeInformation, the file also needs to describe the master
+key that was used to encrypt the local keys. The master keys are
+described by name, their version, and the encryption algorithm.
+
+```
+message EncryptionKey {
+  optional string keyName = 1;
+  optional uint32 keyVersion = 2;
+  optional EncryptionAlgorithm algorithm = 3;
+}
+```
+
+The encryption algorithm is stored using an enumeration and since
+ProtoBuf uses the 0 value as a default, we added an unused value. That
+ensures that if we add a new algorithm that old readers will get
+UNKNOWN_ENCRYPTION instead of a real value.
+
+```
+enum EncryptionAlgorithm {
+  // used for detecting future algorithms
+  UNKNOWN_ENCRYPTION = 0;
+  // 128 bit AES/CTR
+  AES_128 = 1;
+  // 256 bit AES/CTR
+  AES_256 = 2;
+}
+```
+
+## Encryption Variants
+
+Each encrypted column is written as two variants:
+
+* encrypted unmasked - for users with access to the key
+* unencrypted masked - for all other users
+
+The changes to the format were done so that old ORC readers will read
+the masked unencrypted data. Encryption variants encrypt a subtree of
+columns and use a single local key. The initial version of encryption
+support only allows the two variants, but this may be extended later
+and thus readers should use the first variant of a column that the
+reader has access to.
+
+```
+message EncryptionVariant {
+  // the column id of the root
+  optional uint32 root = 1;
```

Comment:
What is the root used for?

Reply:
Root is the column id of that variant.

Reply:
Given a type tree, this is the root of a subtree that is encrypted by a local key.